### PR TITLE
finalized base workchains with examples

### PR DESCRIPTION
**Pull request by pzarabadip**
Hi @yakutovicha 
This pull request covers addition of `aiida_base` workchains which covers three types of `RASPA` calculations. 
It also contains addition of `multiply_untilcell` function to plugin `utils`.

### CODE REVIEW

File: aiida_raspa/workchains/aiida_base.py [hunk 1]
```diff
@@ -0,0 +1,437 @@
+"""
```
Comment (yakutovicha) [inline aiida_raspa/workchains/aiida_base.py]: So the idea behind the usage of this [`aiida_base` workchain](https://github.com/aiidateam/aiida-cp2k/blob/develop/aiida_cp2k/workchains/aiida_base_restart.py)  that we keep it unchanged. The usage of it allowed to make such a small [`Cp2kBaseWorkChain`](https://github.com/aiidateam/aiida-cp2k/blob/develop/aiida_cp2k/workchains/base.py). So my suggestion would be to copy `aiida_base` entirely and then to create a new file with much smaller `RaspaBaseWorkChain` in it.

File: aiida_raspa/workchains/aiida_base.py [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+        if isinstance(port, PortNamespace):
+            wrapped[key] = wrap_bare_dict_inputs(port, value)
+        elif port.valid_type == Dict and isinstance(value, dict):
+            wrapped[key] = Dict(dict=value)
+        else:
+            wrapped[key] = value
+
+    return wrapped
+
+def check_widom_convergence(output_widom=None, comp_list=None, structure_label=None, conv_threshold=None):
```
Comment (yakutovicha) [inline aiida_raspa/workchains/aiida_base.py]: I would suggest to put it into utils, same for `check_gcmc_convergence`, `check_gemc_convergence`, `check_gemc_box`.

File: aiida_raspa/workchains/aiida_base.py [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+            else:
+                box_two_stat.append(False)
+            if output_gemc["box_two"]["general"]["box_cz_average"] > 2*cutoff:
+                box_two_stat.append(True)
+            else:
+                box_two_stat.append(False)
+
+    return box_one_stat, box_two_stat
+
+class RaspaBaseWorkChain(WorkChain):
```
Comment (yakutovicha) [inline aiida_raspa/workchains/aiida_base.py]: In principle the idea behind the `...BaseWorkChain` that it resolve general problems: time exceeded, node failure and etc.. For more specific jobs one should probably create an upper-level workchain that calls the `Base`

File: aiida_raspa/workchains/gemc_workchain.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
+"""
+RaspaGEMCWorkChain : Rapspa workchain for GEMC calculation through RaspaBaseWorkChain
```
[the remark this answers is not in the document]
Comment (pzarabadip) [reply aiida_raspa/workchains/gemc_workchain.py]: Done! These sub-modules are removed in the current version.

File: examples/workchains/test_widom_workchain_2comp.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+                    "WidomProbability": 1.0,
+                    "CreateNumberOfMolecules": 0,
+                    "BlockPocketsFileName": "block_tcc1rs_xenon",
+                },
+            },
+        })
+
+    # framework
+    pwd = os.path.dirname(os.path.realpath(__file__))
+    structure = CifData(file=os.path.join(pwd, '../simple_calculations/files', 'TCC1RS.cif'))
```
[the remark this answers is not in the document]
Comment (pzarabadip) [reply examples/workchains/test_widom_workchain_2comp.py]: Done!

File: aiida_raspa/workchains/gcmc_workchain.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    def define(cls, spec):
+        """
+        Defining the class inputs.
+        """
+        super(RaspaGCMCWorkChain, cls).define(spec)
+
+        spec.expose_inputs(RaspaBaseWorkChain, namespace='raspa_base')
+
+        spec.input('conv_threshold', valid_type=Float, required=True)
+        spec.input('additional_cycle', valid_type=Int, required=True)
```
[the remark this answers is not in the document]
Comment (pzarabadip) [reply aiida_raspa/workchains/gcmc_workchain.py]: Done!

File: aiida_raspa/workchains/gemc_workchain.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+            length_string = self.ctx.inputs.raspa.parameters.get_dict()["System"]["box_one"]["BoxLengths"]
+            box_length = []
+            for element in length_string.split():
+                try:
+                    box_length.append(float(element))
+                except ValueError:
+                    pass
+            # TODO: Make the box addition value smart. #pylint: disable=fixme
+            # TODO: It can be taken from difference of initital and final size. #pylint: disable=fixme
```
[the remark this answers is not in the document]
Comment (pzarabadip) [reply aiida_raspa/workchains/gemc_workchain.py]: Done! These parts have been moved to inspection tools and taken care off.

File: examples/simple_calculations/test_base.py [hunk 8]
```diff
@@ -11,7 +11,7 @@
 from aiida.engine import run_get_pk, run
 from aiida.orm import Code, Dict
 from aiida.plugins import DataFactory
-from aiida_raspa.calculations import RaspaCalculation
+from aiida_raspa.calculations import RaspaCalculation  #pylint: disable=unused-import
```
[the remark this answers is not in the document]
Comment (pzarabadip) [reply examples/simple_calculations/test_base.py]: True. I was mistaken as the builder takes care of it. These imports are removed from all examples in `simple_calculations` folder. Done!

File: examples/simple_calculations/test_binary_mixture.py [hunk 9]
```diff
@@ -9,7 +9,7 @@
 from aiida.common import NotExistent
 from aiida.engine import run_get_pk, run
 from aiida.orm import Code, Dict
-from aiida_raspa.calculations import RaspaCalculation
+from aiida_raspa.calculations import RaspaCalculation  # pylint: disable=unused-import
```
[the remark this answers is not in the document]
Comment (pzarabadip) [reply examples/simple_calculations/test_binary_mixture.py]: Done!

File: examples/simple_calculations/test_base.py [hunk 10]
```diff
@@ -64,41 +64,34 @@ def main(codelabel, submit):
...
-            "store_provenance": True,
-        }
-    }
+    builder.metadata.dry_run = False
+    builder.metadata.store_provenance = True
 
     if submit:
         print("Testing RASPA with simple input ...")
-        res, pk = run_get_pk(RaspaCalculation, **inputs)
+        res, pk = run_get_pk(builder)
```
[the remark this answers is not in the document]
Comment (pzarabadip) [reply examples/simple_calculations/test_base.py]: Yes it is :)

File: aiida_raspa/workchains/gcmc_workchain.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+FolderData = DataFactory('folder')  #pylint: disable=invalid-name
+RaspaCalculation = CalculationFactory("raspa")  #pylint: disable=invalid-name
+
+
+class RaspaGCMCWorkChain(WorkChain):
+    """
+    Handling the GCMC calculation through RaspaBaseWorkChain
+    """
+
+    _calculation_class = RaspaCalculation
```
[the remark this answers is not in the document]
Comment (pzarabadip) [reply aiida_raspa/workchains/gcmc_workchain.py]: Done!

File: aiida_raspa/workchains/gcmc_workchain.py [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+from aiida.orm import Dict, Float, Int
+from aiida.engine import submit  #pylint: disable=unused-import
+from aiida.engine import ToContext, WorkChain, while_
+from aiida_raspa.workchains import RaspaBaseWorkChain
+
+from aiida_raspa.utils.multiply_unitcell import multiply_unit_cell
+from aiida_raspa.utils.inspection_tools import check_gcmc_convergence
+
+ParameterData = DataFactory("dict")  #pylint: disable=invalid-name
+FolderData = DataFactory('folder')  #pylint: disable=invalid-name
```
[the remark this answers is not in the document]
Comment (pzarabadip) [reply aiida_raspa/workchains/gcmc_workchain.py]: Nice point. Although this part of the code is removed, I revised the similar cases in workchain example files, if we would be satisfied, I can do it for the other examples too.

**Comment by yakutovicha** [comment]
Hi @pzarabadip. I fixed things, please have a careful look at my modifications. 

If you are fine with them, I will merge the PR.

Nothing else needs to be done, I think

**Pull request merged**